src/learner.cc

```diff
@@ -345,141 +481,148 @@ class LearnerImpl : public Learner {
     }
   }
 
-  void CheckDataSplitMode() {
-    if (rabit::IsDistributed()) {
-      CHECK(tparam_.dsplit != DataSplitMode::kAuto)
-        << "Precondition violated; dsplit cannot be 'auto' in distributed mode";
-      if (tparam_.dsplit == DataSplitMode::kCol) {
-        // 'distcol' updater hidden until it becomes functional again
-        // See discussion at https://github.com/dmlc/xgboost/issues/1832
-        LOG(FATAL) << "Column-wise data split is currently not supported.";
+  void ConfigureNumFeatures(LearnerAttributes* attr) {
+    // estimate feature bound
+    // TODO(hcho3): Change num_feature to 64-bit integer
+    unsigned num_feature = 0;
+    for (auto & matrix : attr->cache.Container()) {
+      CHECK(matrix.first);
+      CHECK(!matrix.second.ref.expired());
+      const uint64_t num_col = matrix.first->Info().num_col_;
+      CHECK_LE(num_col, static_cast<uint64_t>(std::numeric_limits<unsigned>::max()))
+          << "Unfortunately, XGBoost does not support data matrices with "
+          << std::numeric_limits<unsigned>::max() << " features or greater";
+      num_feature = std::max(num_feature, static_cast<uint32_t>(num_col));
+    }
+    // run allreduce on num_feature to find the maximum value
+    rabit::Allreduce<rabit::op::Max>(&num_feature, 1, nullptr, nullptr, "num_feature");
+    if (num_feature > attr->mparam.num_feature) {
+      attr->mparam.num_feature = num_feature;
+    }
+    CHECK_NE(attr->mparam.num_feature, 0)
+        << "0 feature is supplied.  Are you using raw Booster interface?";
+    // Remove these once binary IO is gone.
+    attr->cfg["num_feature"] = common::ToString(attr->mparam.num_feature);
+    attr->cfg["num_class"] = common::ToString(attr->mparam.num_class);
+  }
+
+  void ConfigureGBM(LearnerAttributes *attr,
+                    LearnerTrainParam const &old, Args const &args) const {
+    if (attr->gbm == nullptr || old.booster != attr->tparam.booster) {
```

Comment:
If Learner had access to parameters on its construction, we could get rid of these chains of Configure methods.

Reply:
It doesn't.  The parameters can be loaded from pickle.